src/output.c

```diff
@@ -113,10 +116,21 @@ output_frame_notify(struct wl_listener *listener, void *data)
 		output_apply_gamma(output);
 	} else {
 		output->pending.tearing_page_flip =
-			get_tearing_preference(output);
-
-		lab_wlr_scene_output_commit(output->scene_output,
-			&output->pending);
+			get_tearing_preference(output) &&
+			output->tearing_fail_count < TEARING_FAIL_LIMIT;
+
+		if (!lab_wlr_scene_output_commit(output->scene_output,
+				&output->pending) && output->pending.tearing_page_flip) {
+			if (++output->tearing_fail_count >= TEARING_FAIL_LIMIT) {
+				wlr_log(WLR_INFO, "Tearing failed for %u consecutive"
+					" frames, disabling", (unsigned int)TEARING_FAIL_LIMIT);
+			}
+			output->pending.tearing_page_flip = false;
+			lab_wlr_scene_output_commit(output->scene_output,
+				&output->pending);
+		} else if (output->pending.tearing_page_flip) {
+			output->tearing_fail_count = 0;
+		}
```

Comment:
I think restructing this makes the logic a bit easier to read through (the setting of `scene_output` could be moved further up and used in the sanity check at the top of the function):
```suggestion
		struct wlr_scene_output *scene_output = output->scene_output;
		struct wlr_output_state *pending = &output->pending;
		
		bool committed =
			lab_wlr_scene_output_commit(scene_output, pending);
		
		if (!pending->tearing_page_flip) {
			return;
		}
		
		if (committed) {
			output->tearing_fail_count = 0;
		} else {
			if (++output->tearing_fail_count >= TEARING_FAIL_LIMIT) {
				wlr_log(WLR_INFO, "setting tearing allowance failed "
					"for %u consecutive frames, disabling",
					(unsigned int)TEARING_FAIL_LIMIT);
			}
			pending->tearing_page_flip = false;
			lab_wlr_scene_output_commit(scene_output, pending);
		}
```